### DISCUSSION TITLE
is there any way to place the line types in the bar so that I can see the focus when I use a cycle of lines assigned to a button?

### DISCUSSION THREAD

**Discussion by vladQll** [xournalpp/xournalpp#4564] [Q&A]
previously I had the geometric shapes inside the floating box so when I pressed the button and I didn't know exactly if I was selecting the right tool, that's why when I put it on the bar it shows me the focus and that's great.

the inconvenient is with the type of line since there are 4 and when I press the button I don't know exactly if I chose the type desired line and I can't take these elements out and place them in the bar to see the focus (which would be very useful) because these elements are like a combo inside the pencil so I have to press several times and try to choose the right one.
standard
dashed
dashed/stippled
dotted

is there any way to place the line types in the bar so that I can see the focus when I use a cycle of lines assigned to a button?

![image](https://user-images.githubusercontent.com/28522011/210758373-3b668fb1-3b59-47cf-b232-a86d1fd8f5bd.png)

this is my configuration of the buttons in the main of the shorcutsPlugin plugin:
![image](https://user-images.githubusercontent.com/28522011/210761769-ef17348c-022a-4f32-8053-393dfed18dfd.png)

**Comment by rolandlo** [comment]
Currently it's not possible, but it would be easy to add.

At the end of `ToolMenuHandler::initToolItems()` (e.g. line 570 of `src/core/gui/toolmenubar/ToolMenuHandler.cpp`, just before connecting the glade signals) add lines like this one (for dotted line style):

```c++    
addCustomItemTgl("DOTTED", ACTION_TOOL_LINE_STYLE_DOT, GROUP_LINE_STYLE, true, "line-style-dot", _("dotted"));
```
You get the names from looking at `ToolMenuHandler::initPenToolItem() `.
Similarly for eraser types.

Then you can use it in a customized toolbar (menu `View > Toolbars > Customize`).

@vladQll Do you know how to compile from source and create a pull request?

**Reply by vladQll** [reply]
i'm not sure what a pull request is, and about compiling from source, well i think that must be similar to installing a program on linux (the latter) if i have done it because i try to follow what the directions say about installing .rpms on fedora, ./configure, make etc, similar things. if that's what you mean by compiling from source i think i could do it. 
in fact i am not a professional programmer but i understand basic coding things. that's why i could understand a little bit the code in the image above but i will try to look for what you mentioned in line 570, by the way right now i am in windows 11

I apologize if what I am writing is not understood my language is Spanish and I am using deepL as a translator.

**Reply by vladQll** [reply]
> Actualmente no es posible, pero sería fácil de agregar.
> 
> Al final de `ToolMenuHandler::initToolItems()`(por ejemplo, la línea 570 de `src/core/gui/toolmenubar/ToolMenuHandler.cpp`, justo antes de conectar las señales del claro) agregue líneas como esta (para el estilo de línea punteada):
> 
> ```c++
> addCustomItemTgl("DOTTED", ACTION_TOOL_LINE_STYLE_DOT, GROUP_LINE_STYLE, true, "line-style-dot", _("dotted"));
> ```
> 
> Obtienes los nombres al mirar `ToolMenuHandler::initPenToolItem() `. Del mismo modo para los tipos de borrador.
> 
> Luego puede usarlo en una barra de herramientas personalizada (menú `View > Toolbars > Customize`).
> 
> @vladQll¿Sabe cómo compilar desde la fuente y crear una solicitud de extracción?

hello so i need to make four of those if i want it to show me something similar to what i want to get would the following be ok?

`addCustomItemTgl("DOTTED", ACTION_TOOL_LINE_STYLE_DOT, GROUP_LINE_STYLE, true, "line-style-dot", _("dotted"));`
`addCustomItemTgl("DASHED", ACTION_TOOL_LINE_STYLE_DOT, GROUP_LINE_STYLE, true, "line-style-dash", _("dashed"));`
`addCustomItemTgl("DASH-/ DOTTED", ACTION_TOOL_LINE_STYLE_DOT, GROUP_LINE_STYLE, true, "line-style-dash-dot", _("dash-/ dotted"));`
`addCustomItemTgl("STANDARD", ACTION_TOOL_LINE_STYLE_DOT, GROUP_LINE_STYLE, true, "line-style-plain", _("standard"));`

**Reply by rolandlo** [reply]
Yes, that looks right.

**Reply by rolandlo** [reply]
> i'm not sure what a pull request is, and about compiling from source, well i think that must be similar to installing a program on linux (the latter) if i have done it because i try to follow what the directions say about installing .rpms on fedora, ./configure, make etc, similar things. if that's what you mean by compiling from source i think i could do it. in fact i am not a professional programmer but i understand basic coding things. that's why i could understand a little bit the code in the image above but i will try to look for what you mentioned in line 570, by the way right now i am in windows 11
> 
> I apologize if what I am writing is not understood my language is Spanish and I am using deepL as a translator.

Compiling from source is described in the Wiki: [Linux](https://github.com/xournalpp/xournalpp/blob/master/readme/LinuxBuild.md), [Windows](https://github.com/xournalpp/xournalpp/blob/master/readme/WindowsBuild.md). It's easier on Linux.

A pull request is the request to get your code changes into the official version.
For creating a pull request you would first have to fork the repo, clone your fork with git, apply your changes, commit them, push them, click on "Create Pull Request" on [GitHub](https://github.com/xournalpp/xournalpp/pulls) and create it.
It's a bit complicated if you haven't used Git before, so you could also post the modified file (ToolMenuHandler.cpp) here and I will create the pull request for you.

**Reply by vladQll** [reply]
thank you very much for the answer my question is the following once this is done I should uninstall my installed xournal and compile or build from here: https://github.com/xournalpp/xournalpp/blob/master/readme/WindowsBuild.md

i have xournal installed on windows 11

**Reply by rolandlo** [reply]
In the end you can install the improved version you built, but it is not necessary. You can also just run from the `build` folder (typing `./xournalpp.exe` in the MinGW64 console) which gets created when building the improved version.

**Reply by vladQll** [reply]
hello i didn't understand very well you told me that i don't need to install the version i just edited? and that i can just run it from the build folder? where is that build folder? so far i have done the following:

1. clone the project in my download folder
2. edit and add in line 570 of the mentioned file `ToolMenuHandler.cpp` 
3. save the file 
4. so far i have installed xournal 1.1.3+dev , libgtk 3.24.35 on windows11

**Reply by rolandlo** [reply]
Now you need to build from source as described [here](https://github.com/xournalpp/xournalpp/blob/master/readme/Compile.md#compile). You should first install the dependencies though (as described [here](https://github.com/xournalpp/xournalpp/blob/master/readme/WindowsBuild.md#install-dependencies)).

**Reply by vladQll** [reply]
> > No estoy seguro de qué es una solicitud de extracción, y sobre compilar desde la fuente, bueno, creo que debe ser similar a instalar un programa en Linux (el último) si lo he hecho porque trato de seguir lo que dicen las instrucciones sobre la instalación .rpms en fedora, ./configure, make, etc., cosas similares. si eso es lo que quieres decir con compilar desde la fuente, creo que podría hacerlo. de hecho, no soy un programador profesional, pero entiendo cosas básicas de codificación. por eso pude entender un poco el código de la imagen de arriba pero trataré de buscar lo que mencionaste en la línea 570, por cierto ahorita estoy en windows 11
> > Pido disculpas si no se entiende lo que estoy escribiendo mi idioma es español y estoy usando deepL como traductor.
> 
> La compilación desde la fuente se describe en Wiki: [Linux](https://github.com/xournalpp/xournalpp/blob/master/readme/LinuxBuild.md) , [Windows](https://github.com/xournalpp/xournalpp/blob/master/readme/WindowsBuild.md) . Es más fácil en Linux.
> 
> Una solicitud de incorporación de cambios es la solicitud para obtener los cambios de código en la versión oficial. Para crear una solicitud de extracción, primero tendría que bifurcar el repositorio, clonar su bifurcación con git, aplicar sus cambios, confirmarlos, enviarlos, hacer clic en "Crear solicitud de extracción" en [GitHub](https://github.com/xournalpp/xournalpp/pulls) y crearlo. Es un poco complicado si no ha usado Git antes, por lo que también puede publicar el archivo modificado (ToolMenuHandler.cpp) aquí y crearé la solicitud de extracción para usted.

Thank you very much that seems better to me so I will send him the file so he can review it and can generate an extraction request, but I still have a question about that if the request is approved that means that it is added to a new version of xournal? I have to download the new xournal from the site to install it and if that is the case how long does it take?
archivo `ToolMenuHandler.cpp`:
[ToolMenuHandler.zip](https://github.com/xournalpp/xournalpp/files/10352605/ToolMenuHandler.zip)

**Reply by vladQll** [reply]
> Ahora necesita compilar desde el código fuente como se describe [aquí](https://github.com/xournalpp/xournalpp/blob/master/readme/Compile.md#compile) . Sin embargo, primero debe instalar las dependencias (como se describe [aquí](https://github.com/xournalpp/xournalpp/blob/master/readme/WindowsBuild.md#install-dependencies) ).

of course I'll try to give it a try, I hope everything goes well

**Comment by rolandlo** [comment]
When the changes are approved and (usually after 1 day) merged, it will become available in the [nightly build](https://github.com/xournalpp/xournalpp/releases/tag/nightly)
Then you can replace the old version with the nightly build (which you have to download, yes).

**Reply by vladQll** [reply]
thank you very much then I will try to compile manually as to learn something new and if I do not succeed, I would appreciate if you can make the request to extract the ToolMenuHandler.cpp file that I sent in the .zip and already within 1 day or 2 days I will be downloading the new nightly compilation. thank you very much. @rolandlo

**Comment by rolandlo** [comment]
@vladQll Can you please download and install the Windows build from https://github.com/xournalpp/xournalpp/pull/4565#issuecomment-1372369630 and give feedback in #4565 ?

**Reply by vladQll** [reply]
Great bro that's what I wanted , only I have a drawback does not show me the focus when I go through each of the styles , but when I get to the dotted style the focus is on all the elements at once , I send you a video comparing with the focus of the geometric figures something like that I would like to have the behavior.

https://user-images.githubusercontent.com/28522011/210869001-9bfab198-913a-4105-bfff-dc9205963873.mp4

**Reply by rolandlo** [reply]
Ok, there was a mistake with the action names. Easy to fix. New installers are generated in https://github.com/xournalpp/xournalpp/pull/4565#issuecomment-1372677060

**Reply by vladQll** [reply]
> Ok, hubo un error con los nombres de las acciones. Fácil de arreglar. Los nuevos instaladores se generan en [el #4565 (comentario)](https://github.com/xournalpp/xournalpp/pull/4565#issuecomment-1372677060)

thanks @rolandlo  install   https://github.com/xournalpp/xournalpp/pull/4565#issuecomment-1372700371  and I made the necessary configurations and it is perfect now if it shows me the focus I appreciate this greatly as it will help me in my workflow.

I also have a question if it is not too much trouble of course I have seen that the default tool you can choose a pencil and a color (that makes sense) but do you think there is a way to adjust it in the main.lua code of shortcutsPlugin so that for example if I am writing with the blue color and then I draw arrows, lines, circles etc and I need to return to the normal pencil and with flat or standard line style I can return but with the last color I have been using at the moment (blue or any other).

i have configured a button to be the default tool and a highlighter, and as the default tool is the black pencil, every time i am drawing circles lines etc and i try to go back to the pencil it goes back to the default (black).

I have imagined that there can be two codes that I do not know for possible solutions

1) some code that allows to tell the default tool that when it returns to the pencil it returns with the recently used color (I'm not sure but it seems difficult or not?)

2) I think another option would be a code that when I am writing with the color I want and if I am going to use arrows, lines etc. when I get to the last tool in the list of figures and I press the button once more instead of returning to the first figure. lose the focus so that it would be as if I had deselected it and return to the normal pencil.

attached images of the code I have at the moment

![image](https://user-images.githubusercontent.com/28522011/210894045-ed75aba8-a63f-4e19-bcbf-bbdf2a7a8259.png)
 
and here are images of the functions that do the job
![image](https://user-images.githubusercontent.com/28522011/210894172-dc397d51-e2df-4cdc-9d97-dab329f417b9.png)

**Reply by vladQll** [reply]
hello @rolandlo after a while I stopped using xournal for work reasons, I came back and bought a graphic tablet better than the wacom, is a huion q630M or known as inspiroy dial 2, I do not know if you remember but I think that giving a look you will remember what you helped me solve here, my question is the following in this new version of xournal 1.2.0 +dev included this functionality because as I said I just opened the tablet and I will begin to configure my framework, if any problem arises with the buttons I will be consulting later for now I am interested in this functionality.

**Comment by rolandlo** [comment]
This is getting off topic. It would be better to start a new discussion. But briefly: With a plugin you could have a keyboard shortcut that would
1) query the current color (via `app.getToolInfo("active")["color"]`) and save it in some variable, then
2) switch to the default tool
3) switch to the saved color

It won't work from the toolbar though, just with a keyboard shortcut.
